homeassistant/components/sonos/__init__.py

```diff
@@ -184,21 +188,63 @@ def is_device_invisible(self, ip_address: str) -> bool:
         """Check if device at provided IP is known to be invisible."""
         return any(x for x in self._known_invisible if x.ip_address == ip_address)
 
-    def _create_visible_speakers(self, ip_address: str) -> None:
-        """Create all visible SonosSpeaker instances with the provided seed IP."""
-        try:
-            soco = SoCo(ip_address)
+    async def async_subscribe_to_zone_updates(self, ip_address: str) -> None:
+        """Test subscriptions and create SonosSpeakers based on results."""
+        soco = SoCo(ip_address)
+        # Cache now to avoid household ID lookup during first ZoneGroupState processing
+        await self.hass.async_add_executor_job(
+            getattr,
+            soco,
+            "household_id",
+        )
+        sub = await soco.zoneGroupTopology.subscribe()
+
+        @callback
+        def _async_add_visible_zones(subscription_succeeded: bool = False) -> None:
+            """Determine visible zones and create SonosSpeaker instances."""
+            subscription = None
             visible_zones = soco.visible_zones
             self._known_invisible = soco.all_zones - visible_zones
-        except (OSError, SoCoException) as ex:
+            for zone in visible_zones:
+                if zone.uid not in self.data.discovered:
+                    if subscription_succeeded and zone is soco:
+                        subscription = sub
+                    self.hass.async_create_task(
+                        self.async_add_speaker(zone, subscription)
+                    )
```

Comment:
Would it make sense to have an `async_add_speakers` so you don't have a thundering heard waiting in line for the lock?